src/environmentd/tests/sql.rs

```diff
@@ -1291,6 +1291,132 @@ fn test_explain_timestamp_json() {
     let _explain: TimestampExplanation<Timestamp> = serde_json::from_str(&explain).unwrap();
 }
 
+// Verify that `EXPLAIN TIMESTAMP ...` within acts like a peek within a transaction.
+// That is, ensure the following:
+// 1. Consistently returns its transaction timestamp as the "query timestamp"
+// 2. Acquires read holds for all objects within the same time domain
+// 3. Errors during a write-only transaction
+// 4. Errors when an object outside the chosen time domain is referenced
+#[test]
+fn test_github_18950() {
+    // Set the timestamp to zero for deterministic initial timestamps.
+    let nowfn = Arc::new(Mutex::new(NOW_ZERO.clone()));
+    let now = {
+        let nowfn = Arc::clone(&nowfn);
+        NowFn::from(move || (nowfn.lock().unwrap())())
+    };
+
+    let config = util::Config::default()
+        .workers(2)
+        .with_now(now)
+        .unsafe_mode();
+
+    let server = util::start_server(config).unwrap();
+
+    let mut client_writes = server.connect(postgres::NoTls).unwrap();
+    let mut client_reads = server.connect(postgres::NoTls).unwrap();
+
+    client_writes
+        .batch_execute("CREATE TABLE t1 (i1 int)")
+        .unwrap();
+
+    // Verify execution during a write-only txn fails
+    client_writes.batch_execute("BEGIN").unwrap();
+    client_writes
+        .batch_execute("INSERT INTO t1 VALUES (1)")
+        .unwrap();
+    let error = client_writes
+        .query_one("EXPLAIN TIMESTAMP FOR SELECT * FROM t1;", &[])
+        .unwrap_err();
+
+    assert!(format!("{}", error).contains("transaction in write-only mode"));
+
+    client_writes.batch_execute("ROLLBACK").unwrap();
+
+    // Verify the transaction timestamp is returned for each select and
+    // read holds are acquired as of the transaction timestamp
+    client_reads.batch_execute("BEGIN").unwrap();
+    let mut query_timestamp = None;
+    let mut t1_read_frontier = None;
+
+    for i in 1..5 {
+        let row = client_reads
+            .query_one("EXPLAIN TIMESTAMP AS JSON FOR SELECT * FROM t1;", &[])
+            .unwrap();
+
+        let explain: String = row.get(0);
+        let explain: TimestampExplanation<Timestamp> = serde_json::from_str(&explain).unwrap();
+        let explain_timestamp = explain.determination.timestamp_context.timestamp().unwrap();
+
+        if let Some(timestamp) = query_timestamp {
+            assert_eq!(timestamp, *explain_timestamp);
+        } else {
+            query_timestamp = Some(*explain_timestamp);
+        }
+
+        let explain_t1_read_frontier = explain
+            .sources
+            .first()
+            .unwrap()
+            .read_frontier
+            .first()
+            .unwrap();
+
+        // Ensure `t1` does not undergo compaction
+        if let Some(timestamp) = t1_read_frontier {
+            assert_eq!(timestamp, *explain_t1_read_frontier);
+        } else {
+            t1_read_frontier = Some(explain_t1_read_frontier.clone());
+        }
```

Comment:
Can you explain why? My thinking here is we should be testing that the read frontier of the source isn't advancing

Reply:
If the read frontier started as less than the query timestamp, then it can advance up to the query timestamp. So maybe a better assertion would be that the read frontier remains less than or equal to the query timestamp.

For a concrete example, let's say `t` has a read frontier of `10` and the query timestamp is `20`. Then the read frontier can advance up to `20`.

Reply:
Ahh right, that makes sense. Updated